In [11]:
from copy import deepcopy
import time
def doc_calc(table , x1 , y1):
    if table[x1][y1] == "red":
        return(1)
    else:
        return(0)
def potion_calc(table , x1 , y1):
    if table[x1][y1] == "green":
        return(1)
    else:
        return(0)
t1 = time.time()
f = open("test1.in" , "r")
a = f.read().split()
potion_x = [] ; potion_y = [] ; double_x = [] ; double_y = [] ; block_x = [] ; block_y = [];queue_x = [0] ; queue_y = [0]; potions = 0;
queue_parent_x = [0]; queue_parent_y = [0]; parent_num = [0]; docs = 1; docs_x = [0]; docs_y = [0]; visited_x =[]; visited_y =[]; state_docs = [1] ; state_potions = []; state_double = [];
n = int(a[0])
m = int(a[1])
c = int(a[2])
k = int(a[3])
j = 4
class state:
    def __init__(self, s_table , s_docs_x , s_docs_y , parent_state,doc_id,moves,potions):
        self.s_table = deepcopy(s_table)
        self.s_docs_x = deepcopy(s_docs_x)
        self.s_docs_y = deepcopy(s_docs_y)
        self.parent_state = deepcopy(parent_state)
        self.doc_id = deepcopy(doc_id)
        self.moves = deepcopy(moves)
        self.potions = deepcopy(potions)


state_potions.append(c)
table = [[0] * m for i in range(n)]
for i in range(n):
    for p in range(m):
        table[i][p]= "white"
table[n-1][m-1] = 'orange'
for i in range(c):
    potion_x.append(int(a[j]))
    potion_y.append(int(a[j+1]))
    table [int(a[j])][int(a[j+1])] = "green"
    j = j + 2
for i in range(k):
    double_x.append(int(a[j]))
    double_y.append(int(a[j+1]))
    table [int(a[j])][int(a[j+1])] = "red"
    j = j + 2
d =  int(a[j])
j = j + 1
for i in range(d):
    block_x.append(int(a[j]))
    block_y.append(int(a[j+1]))
    table [int(a[j])][int(a[j+1])] = "black"
    j = j + 2
f = 0
states = []
docs_x2 = [0 , n-1]
docs_y2 = [0 , 0]
if table[0][0] != "red":
    states.append(state(table , docs_x,docs_y , 0 , 0 , 0 , 0))
else:
    states.append(state(table , docs_x2,docs_y2 , 0 , 0 , 0 , 0))

while queue_x : 
   
    cnt = 0
    end = 0
    for i1 in states[f].s_docs_x:
        if not(states[f].s_docs_x[cnt] == n-1 and states[f].s_docs_y[cnt] == m-1):
            end = 1
        #cnt = cnt + 1
    if end == 0 and states[f].potions == c:
        print("done" , states[f].moves - 1)
        ans = f
        break
    x1 =queue_x.pop(0)
    y1 =queue_y.pop(0)
  #  print(queue_x , queue_y)
    states[f].s_docs_x[states[f].doc_id]=x1
    states[f].s_docs_y[states[f].doc_id]=y1
    if states[f].s_table[x1][y1] == "green":
        states[f].s_table[x1][y1] = "white"
        states[f].potions = states[f].potions + 1
    if states[f].s_table[x1][y1] == "red":
        states[f].s_docs_x.append(n-1)
        states[f].s_docs_y.append(0)
        states[f].s_table[x1][y1] = "white"
    l = 0
    for i in states[f].s_docs_x:
        xd = states[f].s_docs_x[l]
        yd = states[f].s_docs_y[l]
       
        check = 0
        v = 0
        for q in states:
            if q.s_table == states[f].s_table and q.s_docs_x == states[f].s_docs_x and q.s_docs_y == states[f].s_docs_y and v != f:
                check = 1
            v = v + 1
       # if f==1:
        #    print(states[0].s_docs_x , states[1].s_docs_x )
        if xd < n - 1 and states[f].s_table[xd+1][yd] != "black" and not (check == 1 ):
            queue_x.append(xd+1)
            queue_y.append(yd)
            states.append(state(states[f].s_table , states[f].s_docs_x , states[f].s_docs_y , f ,l , states[f].moves + 1 , states[f].potions))
        if xd > 0 and states[f].s_table[xd-1][yd] != "black" and not (check == 1 ):
            queue_x.append(xd-1)
            queue_y.append(yd)
            states.append(state(states[f].s_table , states[f].s_docs_x , states[f].s_docs_y , f ,l , states[f].moves + 1 , states[f].potions))
        if yd > 0 and states[f].s_table[xd][yd-1] != "black" and not (check == 1 ):
            queue_x.append(xd)
            queue_y.append(yd-1)
            states.append(state(states[f].s_table , states[f].s_docs_x , states[f].s_docs_y , f ,l , states[f].moves + 1 , states[f].potions))
        if yd < m-1 and states[f].s_table[xd][yd+1] != "black" and not (check == 1 ):
            queue_x.append(xd)
            queue_y.append(yd+1)
            states.append(state(states[f].s_table , states[f].s_docs_x , states[f].s_docs_y , f ,l , states[f].moves + 1 ,states[f].potions ))
        l = l + 1
    f = f + 1
t2 = time.time()
z1 = ans
cnt2 = states[z1].moves
while cnt2 > 0:
    print(list(zip(states[z1].s_docs_x,states[z1].s_docs_y)))
    z1 = states[z1].parent_state
    cnt2 = cnt2 - 1
print(t2 - t1)

done 10
[(3, 3), (3, 0)]
[(3, 3), (3, 0)]
[(2, 3), (3, 0)]
[(1, 3), (3, 0)]
[(0, 3), (3, 0)]
[(0, 2), (3, 0)]
[(0, 1), (3, 0)]
[(0, 0)]
[(1, 0)]
[(2, 0)]
[(1, 0)]
0.021963119506835938


In [16]:
from copy import deepcopy
import time

ch = 0
done = 0
explored = dict()
def print_path(states , f):
    cnt = states[f].moves
    ans = f
    while cnt > 0:
        print(list(zip(states[ans].s_docs_x , states[ans].s_docs_y)))
        ans = states[ans].parent_state
        cnt = cnt - 1
    

def dfs(states , len_n , max_len , f):
    global done
    if done == 1:
        return
    if (len_n) > max_len:
        return
    global t1
    global ch 
    ch += 1
    end = 0
    for i in range(len(states[f].s_docs_x)):
        if not (states[f].s_docs_x[i] == n-1 and states[f].s_docs_y[i] == m-1):
            end = 1
        
    if end == 0 and states[f].potions == c:
        print("done" , states[f].moves )
        print_path(states , f)
        ans = f
        done = 1
        t2 = time.time()
        print(t2 - t1)
        return ans 
        
    check = 0
    if states[f].hs() in explored:
        if states[f].ls >= explored[states[f].hs()]:
            check = 1

    if check != 1:     
        explored[states[f].hs()] = states[f].ls
        for l in range(len(states[f].s_docs_x)):
            xd = states[f].s_docs_x[l]
            yd = states[f].s_docs_y[l]

            if xd < n - 1 and states[f].s_table[xd+1][yd] != "black":
                states.append(state(states[f].s_table , states[f].s_docs_x , states[f].s_docs_y , f ,l , states[f].moves + 1 , states[f].potions , len_n + 1))
                size = len(states)-1
                states[size].s_docs_x[l] = xd + 1
                states[size].s_docs_y[l] = yd
                if states[f].s_table[xd+1][yd] == "green":
                    states[size].s_table[xd+1][yd] = "white"
                    states[size].potions = states[f].potions + 1
                if states[f].s_table[xd+1][yd] == "red":
                    states[size].s_docs_x.append(n-1)
                    states[size].s_docs_y.append(0)
                    states[size].s_table[xd+1][yd] = "white"
                dfs (states , len_n + 1 , max_len , size )

            if xd > 0 and states[f].s_table[xd-1][yd] != "black":
                states.append(state(states[f].s_table , states[f].s_docs_x , states[f].s_docs_y , f ,l , states[f].moves + 1 , states[f].potions, len_n + 1))
                size = len(states)-1
                states[size].s_docs_x[l] = xd - 1
                states[size].s_docs_y[l] = yd
                if states[f].s_table[xd-1][yd] == "green":
                    states[size].s_table[xd-1][yd] = "white"
                    states[size].potions = states[f].potions + 1
                if states[f].s_table[xd-1][yd] == "red":
                    states[size].s_docs_x.append(n-1)
                    states[size].s_docs_y.append(0)
                    states[size].s_table[xd-1][yd] = "white"
                dfs (states , len_n + 1 , max_len , size )

            if yd > 0 and states[f].s_table[xd][yd-1] != "black":
                states.append(state(states[f].s_table , states[f].s_docs_x , states[f].s_docs_y , f ,l , states[f].moves + 1 , states[f].potions , len_n + 1))
                size = len(states)-1
                states[size].s_docs_x[l] = xd 
                states[size].s_docs_y[l] = yd - 1
                if states[f].s_table[xd][yd-1] == "green":
                    states[size].s_table[xd][yd-1] = "white"
                    states[size].potions = states[f].potions + 1
                if states[f].s_table[xd][yd-1] == "red":
                    states[size].s_docs_x.append(n-1)
                    states[size].s_docs_y.append(0)
                    states[size].s_table[xd][yd-1] = "white"
                dfs (states , len_n + 1 , max_len , size)

            if yd < m-1 and states[f].s_table[xd][yd+1] != "black":
                states.append(state(states[f].s_table , states[f].s_docs_x , states[f].s_docs_y , f ,l , states[f].moves + 1 ,states[f].potions , len_n + 1))
                size = len(states)-1
                states[size].s_docs_x[l] = xd 
                states[size].s_docs_y[l] = yd +1
                if states[f].s_table[xd][yd+1] == "green":
                    states[size].s_table[xd][yd+1] = "white"
                    states[size].potions = states[f].potions + 1
                if states[f].s_table[xd][yd+1] == "red":
                    states[size].s_docs_x.append(n-1)
                    states[size].s_docs_y.append(0)
                    states[size].s_table[xd][yd+1] = "white"
                dfs (states , len_n + 1 , max_len , size)
                            

if __name__ == "__main__":
    t1 = time.time()
    f = open("test1.in" , "r")
    a = f.read().split()
    potion_x = [] ; potion_y = [] ; double_x = [] ; double_y = [] ; block_x = [] ; block_y = [];queue_x = [0] ; queue_y = [0]; potions = 0;
    queue_parent_x = [0]; queue_parent_y = [0]; parent_num = [0]; docs = 1; docs_x = [0]; docs_y = [0]; visited_x =[]; visited_y =[]; state_docs = [1] ; state_potions = []; state_double = [];
    n = int(a[0])
    m = int(a[1])
    c = int(a[2])
    k = int(a[3])
    j = 4
    class state:
        def __init__(self, s_table , s_docs_x , s_docs_y , parent_state,doc_id,moves,potions,ls):
            self.s_table = [r [:] for r in s_table]
            self.s_docs_x = s_docs_x[:]
            self.s_docs_y = s_docs_y[:]
            self.parent_state = parent_state
            self.doc_id = doc_id
            self.moves = moves
            self.potions = potions
            self.ls = ls
        def __eq__(self, other):
            return self.s_table == other.s_table and self.s_docs_y == other.s_docs_y and self.s_docs_y == other.s_docs_y
        def hs(self):
            return str(self.s_table) + str(self.s_docs_x) + str(self.s_docs_y)

    state_potions.append(c)
    table = [[0] * m for i in range(n)]
    for i in range(n):
        for p in range(m):
            table[i][p]= "white"
    table[n-1][m-1] = 'orange'
    for i in range(c):
        potion_x.append(int(a[j]))
        potion_y.append(int(a[j+1]))
        table [int(a[j])][int(a[j+1])] = "green"
        j = j + 2
    for i in range(k):
        double_x.append(int(a[j]))
        double_y.append(int(a[j+1]))
        table [int(a[j])][int(a[j+1])] = "red"
        j = j + 2
    d =  int(a[j])
    j = j + 1
    for i in range(d):
        block_x.append(int(a[j]))
        block_y.append(int(a[j+1]))
        table [int(a[j])][int(a[j+1])] = "black"
        j = j + 2
    f = 0
    states = []
    docs_x2 = [0 , n-1]
    docs_y2 = [0 , 0]
    if table[0][0] != "red":
        states.append(state(table , docs_x,docs_y , 0 , 0 , 0 , 0 , 0))
    else:
        states.append(state(table , docs_x2,docs_y2 , 0 , 0 , 0 , 0 , 0))
    for i in range(50):
        explored.clear()
        dfs(states.copy() , 0 , i , 0)
        
    print(ch)
   

done 11
[(3, 3), (3, 3)]
[(3, 3), (3, 2)]
[(3, 3), (3, 1)]
[(3, 3), (3, 0)]
[(3, 3), (2, 0)]
[(3, 3), (3, 0)]
[(2, 3), (3, 0)]
[(1, 3), (3, 0)]
[(0, 3), (3, 0)]
[(0, 2), (3, 0)]
[(0, 1), (3, 0)]
0.3580663204193115
23778


My state model include : map , valued by color values of our map (that shows blocked cells , potion cells and double cells)
                        and doctors positions , that we decide by this two the uniqueness of our state 
                        for IDS , we have also the length of our path cause we can get to a state with diffrent number of moves for diffrent lengths
                        
Goal state : We finish our algorithm when all doctors are in goal state and all potions are collected (collected potions number = c)

Initial state : We start our algorithm by a doctor in cell[0][0] , number of potions = 0 and zero moves , with the table made by colors of input

Action : For every doctor in any cell , we can have at most 4 new cells to check , which some might be considered repetitious , and some might be not possible to move into
         so we add the cells that have our conditions to end of our queue
         
Queue : We add every possible move to our queue and do the first move (queue[0]) , then search for new possible moves . Until we see all possible states to get to our goal

BFS : Breadth first search . we add every cell 1 move away from our current cell . It needs less amount of time than IDS and DFS  , But more amount of memory to save our states .

IDS : Iterative deepening search , We start from our first state and max length of 0 , and do DFS till we reach our max lenghth . after we are done with our dfs , we add 
      1 unit to our max length and do DFS again like before . We keep adding max length till we find our path to goal . It needs more time than two others , But less memory than 

Notice that all of our algorithms expect weighted A* find the optimal path .

In terms of memory : A* better than IDS better than BFS .
In terms of time complexity : Weighted A* better than A* better than BFS better than IDS .

Test1:                  

| Algorithm | Path_size | Seen States | Average Time |
| :-: | :-: | :-: | :-: |
| BFS | 11 | 225 | 0.01961 
| IDS | 11 | 23778 | 0.236
| A* | 11 | | 


Test2 :

| Algorithm | Path_size | Seen States | Average Time |
| :-: | :-: | :-: | :-: |
| BFS | 11 | 564 | 0.257
| IDS | 11 | 2585759 | 65.4
| A* | 11 | | 

Test 3 :

| Algorithm | Path_size | Seen States | Average Time |
| :-: | :-: | :-: | :-: |
| BFS | 19 | 7823 | 19.24 |
| IDS | 19 | 2393142 | 68.4
| A* | 19 |

In [18]:
import os

os.system('jupyter nbconvert --to html full_ca.ipynb --output out5')

0